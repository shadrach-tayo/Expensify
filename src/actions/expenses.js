import database from '../firebase/firebase';

// addExpense
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, amount, note, createdAt };
    return database
      .ref('expenses')
      .push(expense)
      .then(ref => {
        dispatch(addExpense({ id: ref.key, ...expense }));
      });
  };
};

// removeExpense
export const removeExpense = ({ id = '' } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// startRemoveExpense
export const startRemoveExpense = ({ id = '' } = {}) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

// editExpense
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// startEditExpense
export const startEditExpense = (id, updates) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .update(updates)
      .then(() => dispatch(editExpense(id, updates)));
  };
};

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref('expenses')
      .once('value')
      .then(snapshots => {
        const expenses = [];
        snapshots.forEach(snapshot => {
          expenses.push({ id: snapshot.key, ...snapshot.val() });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
