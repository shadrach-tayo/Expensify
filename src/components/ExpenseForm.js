import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super();
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    };
  }

  onDescriptionchange = e => {
    this.setState({ description: e.target.value });
  };

  onNoteChange = e => {
    this.setState({ note: e.target.value });
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount: e.target.value });
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      focused
    }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { description, amount, note, createdAt } = this.state;
    if (!description || !amount) {
      this.setState({ error: 'please fill in the description and the amount' });
    } else {
      this.setState({ error: '' });
      this.props.onSubmit({
        amount: parseInt(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        description,
        note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error ? <p>{this.state.error}</p> : ''}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionchange}
          />
          <input
            type="number"
            value={this.state.amount}
            placeholder="amount"
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add expense note (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
