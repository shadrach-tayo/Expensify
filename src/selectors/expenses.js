import moment from 'moment';

// Get Visible Expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment)
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment)
        : true;
      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') return a.createdAt < b.createdAt ? 1 : -1;
      if (sortBy === 'amount') return a.amount < b.amount ? 1 : -1;
    });
};
