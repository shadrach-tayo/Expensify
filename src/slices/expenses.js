import { createSlice } from "@reduxjs/toolkit";

export const initialState = { expenses: [], loading: false };

const expenseSlice = createSlice({
  name: "expenses",
  initialState,
  reducers: {
    addExpense: (state) => {
      state.loading = true;
    },

    setExpense: (state) => ({...state, loading: true}),    

    addExpenseSuccess: (state, { payload }) => {
      return {
        ...state,
        loading: false,
        expenses: [...state.expenses, payload],
      };
    },

    addExpenseError: (state) => {
      state.loading = false;
    },

    setExpensesSuccess: (state, { payload }) => {
      return {
        ...state,
        loading: false,
        expenses: payload,
      };
    },

    removeExpense: (state, { payload }) => {
      return {...state, loading: true}
    },

    removeExpenseSuccess: (state, { payload }) => {
      return {...state, expenses: state.expenses.filter(({ id }) => id !== payload.id)}
    },

    editExpense: (state) => {      
      return {...state, loading: true};
    },
    
    editExpenseSuccess: (state, { payload }) => {      

      const expenses = state.expenses.map((expense) => {
        if (expense.id === payload.id) {
          return {
            ...expense,
            ...payload.updates,
          };
        }
        return expense
      });

      return {...state, expenses};
    },

    setExpense: (_, { payload }) => {
      return payload;
    },
  },
});

export const {
  addExpense,
  removeExpense,
  setExpense,
  editExpense,
  addExpenseSuccess,
  removeExpenseSuccess,
  editExpenseSuccess,
  setExpensesSuccess
} = expenseSlice.actions;

export default expenseSlice.reducer;
