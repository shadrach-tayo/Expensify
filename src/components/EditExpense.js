import React from 'react';
import { connect } from 'react-redux';
import ExpenseFrom from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemove = () => {
    let id = this.props.expense.id;
    this.props.removeExpense({ id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseFrom onSubmit={this.onSubmit} expense={this.props.expense} />
        <button onClick={this.onRemove}>remove item</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: data => dispatch(removeExpense(data))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
