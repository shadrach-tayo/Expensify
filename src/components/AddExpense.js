import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpense extends React.Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h3>Add Expense</h3>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpense);
