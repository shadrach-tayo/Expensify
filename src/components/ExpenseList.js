import React from "react";
import { useSelector } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpense from "../selectors/expenses";
import { getExpenses, getFilters } from "../selectors";

export const ExpenseList = (props) => {
  const { expenses } = useSelector(getExpenses);
  const filters = useSelector(getFilters);

  const visibleExpenses = selectExpense(expenses, filters);
  console.log("visible ", visibleExpenses);

  return (
    <div className="container">
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
      <div className="list-body">
        {visibleExpenses.length === 0 ? (
          <div className="list-item list-item__message">
            <span className="">No expenses</span>
          </div>
        ) : (
          visibleExpenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )}
      </div>
    </div>
  );
};

// const mapStateToProps = (state) => {
//   return {
//     expenses: selectExpense(state.expenses, state.filters),
//   };
// };

export default ExpenseList;
