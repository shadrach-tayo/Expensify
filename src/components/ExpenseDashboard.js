import React, { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getExpenses } from "../selectors";
import { setExpense } from "../slices/expenses";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => {
  const dispatch = useDispatch();
  const _loadExpenses = useCallback(() => dispatch(setExpense(), [dispatch]));

  const { expenses, loading } = useSelector(getExpenses);

  useEffect(() => {
    if(loading) return;
    _loadExpenses();
    return () => {};
  }, []);


  return (
    <div>
      <ExpensesSummary />
      <ExpenseListFilters />
      <ExpenseList />
    </div>
  );
};

export default ExpenseDashboardPage;
