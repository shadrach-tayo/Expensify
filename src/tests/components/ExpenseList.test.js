import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render ExpenseList with empty expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
