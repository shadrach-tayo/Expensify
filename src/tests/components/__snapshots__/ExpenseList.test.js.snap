// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty expenses 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={0}
    createdAt={0}
    description="new expense"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109952}
    createdAt={-345600000}
    description="new expense2"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={500}
    createdAt={345600000}
    description="new expense3"
    id="3"
    key="3"
    note=""
  />
</div>
`;
